### PR TITLE
bug(core): EWAH bitmap in query is very CPU intensive

### PR DESCRIPTION
**Pull request by vishramachandran**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

We use EWAH bitmap in queries. When matching partitions are very high, constructing the bitmap is VERY cpu intensive since the set is O(n) complexity on a bitmap. 

**New behavior :**

Replace bitmap with debox buffers.

Also limiting the maximum number of partition matches per shard in queries to be defensive. Extremely large queries can hog memory and destabilize the server.

### CODE REVIEW

File: core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala [hunk 1]
```diff
@@ -101,8 +102,7 @@ class DownsampledTimeSeriesShard(ref: DatasetRef,
     // Create a ReadablePartition objects that contain the time series data. This can be either a
     // PagedReadablePartitionOnHeap or PagedReadablePartitionOffHeap. This will be garbage collected/freed
     // when query is complete.
-    import filodb.core.Iterators._
-    val partKeys = lookup.partsInMemory.intIterator().map(partKeyFromPartId, 10000) // TODO configure
+    val partKeys = lookup.partsInMemory.iterator().take(10000).map(partKeyFromPartId) // TODO configure
```
Comment (velvia) [inline core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala]: Can we please make it configurable?  Easy to do.

File: core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala [hunk 2]
```diff
@@ -1428,21 +1420,27 @@ class TimeSeriesShard(val ref: DatasetRef,
...
+      if (matches.length > storeConfig.maxQueryMatches)
+        throw new IllegalArgumentException(s"Seeing ${matches.length} matching time series per shard. Try " +
+          s"to narrow your query by adding more filters so there is less than  matches " +
+          s"or request for increasing number of shards this metric lives in")
 
       // first find out which partitions are being queried for data not in memory
-      val _schema = Option(coll.getFirstSetBit).filter(_ >= 0).map(schemaIDFromPartID)
-      val it1 = InMemPartitionIterator(coll.intIterator())
+      val firstPartId = if (matches.isEmpty) None else Some(matches(0))
+      val _schema = firstPartId.filter(_ >= 0).map(schemaIDFromPartID)
```
Comment (velvia) [inline core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala]: Here too, filter not needed anymore

File: core/src/test/scala/filodb.core/memstore/PartKeyLuceneIndexSpec.scala [hunk 3]
```diff
@@ -227,37 +228,37 @@ class PartKeyLuceneIndexSpec extends FunSpec with Matchers with BeforeAndAfter {
...
+//    // Add the first ten keys and row numbers
+//    partKeyFromRecords(dataset6, records(dataset6, readers.take(10)), Some(partBuilder))
+//      .zipWithIndex.foreach { case (addr, i) =>
+//      keyIndex.addPartKey(partKeyOnHeap(dataset6, ZeroPointer, addr), i, System.currentTimeMillis())()
+//    }
+//
+//    keyIndex.refreshReadersBlocking()
+//
+//    val filters1 = Seq(ColumnFilter("Actor2Code", Equals("GOV".utf8)),
+//      ColumnFilter("Actor2Name", Equals("REGIME".utf8)))
```
Comment (velvia) [inline core/src/test/scala/filodb.core/memstore/PartKeyLuceneIndexSpec.scala]: Why is this test commented out?  This seems like an important use case no?

Comment (vishramachandran) [reply core/src/test/scala/filodb.core/memstore/PartKeyLuceneIndexSpec.scala]: good catch... not intended

**Comment by velvia** [comment]
@vishramachandran I forgot to add: could you please add docs/comments to `timeseries-dev-source.conf` for the new param?  Thanks

**Pull request merged**